lightning/src/ln/outbound_payment.rs

```diff
@@ -902,15 +902,64 @@ impl OutboundPayments {
 			}
 		}
 
-		let mut route = router.find_route_with_id(
-			&node_signer.get_node_id(Recipient::Node).unwrap(), &route_params,
+		let payer = node_signer.get_node_id(Recipient::Node).unwrap();
+		let mut route;
+		match router.find_route_with_id(
+			&payer,&route_params,
 			Some(&first_hops.iter().collect::<Vec<_>>()), inflight_htlcs(),
 			payment_hash, payment_id,
-		).map_err(|_| {
-			log_error!(logger, "Failed to find route for payment with id {} and hash {}",
-				payment_id, payment_hash);
-			RetryableSendFailure::RouteNotFound
-		})?;
+		) {
+			Ok(res) => route = res,
+			Err(error) => {
+				if error.err == "Cannot generate a route to ourselves" {
+					let payment_secret = match recipient_onion.payment_secret {
+						Some(secret) => secret,
+						None => PaymentSecret([0; 32])
+					};
+					let payment_preimage = PaymentPreimage([0; 32]);
+					let payment_purpose = PaymentPurpose::InvoicePayment {
+						payment_preimage: Some(payment_preimage),
+						payment_secret,
+					};
+					let dummy_route = Route {
+						paths: vec![Path {
+							hops: vec![RouteHop {
+									pubkey: payer,
+									node_features: NodeFeatures::empty(),
+									short_channel_id: 0,
+									channel_features: ChannelFeatures::empty(),
+									fee_msat: 0,
+									cltv_expiry_delta: 0,
+									maybe_announced_channel: false,
+								}],
+							blinded_tail: None,
+						}],
+						route_params: Some(route_params.clone()),
+					};
+					let _ = self.add_new_pending_payment(payment_hash,
+						recipient_onion.clone(), payment_id, keysend_preimage, &dummy_route, Some(retry_strategy),
+						Some(route_params.payment_params.clone()), entropy_source, best_block_height)
+						.map_err(|_| {
+							log_error!(logger, "Payment with id {} is already pending. New payment had payment hash {}",
+								payment_id, payment_hash);
+							RetryableSendFailure::DuplicatePayment
+						})?;
+					let mut pending_outbounds_lock = self.pending_outbound_payments.lock().unwrap();
+					let payment = pending_outbounds_lock.get_mut(&payment_id).unwrap();
+					payment.mark_fulfilled();
+					let mut pending_events_lock = pending_events.lock().unwrap();
+					pending_events_lock.push_back((Event::PaymentSent { payment_id: Some(payment_id), payment_preimage,
+								payment_hash, fee_paid_msat: None }, None));
+					pending_events_lock.push_back((Event::PaymentClaimable { receiver_node_id: Some(payer), payment_hash,
+								onion_fields: Some(recipient_onion), amount_msat: route_params.final_value_msat, counterparty_skimmed_fee_msat: 0,
+								purpose: payment_purpose, via_channel_id: None, via_user_channel_id: None, claim_deadline: None }, None));
+					return Ok(());
+				}
```

Comment:
I think we should be generating `PaymentClaimed` instead of `PaymentClaimable`. The latter may cause the user to call `claim_funds`, which isn't harmful but is unnecessary IIUC.

Reply:
Good point, will make necessary changes.

Reply:
@G8XSU, you are right. I think adding documentation here will definitely help users.

Reply:
> In general principle of idempotency, self-payments should ideally go through same set of events as a normal payment would. I am not sure but what if user has some extra logic while handling `PaymentClaimable`.
> 
> In any case, if we choose to go this route to directly generate `PaymentClaimed` then we should update the docs on top of it to include this scenario.

I was wondering about this too, like should we fully replicate the steps of a normal payment? That would mean we should generate `PaymentClaimable`, then have logic in `ChannelManager::claim_funds` and `::fail_htlc` to handle this case and generate the final `Payment{Claimed,Failed}` events. Seems like overkill though?